coordiante_transform/src/coordinate_transform_component.cpp

```diff
@@ -0,0 +1,187 @@
+    // iterate over each tag to and transform it into body frame
+    for(int i = 0; i < tag_array_msg->tags.size(); i++) {
+      stsl_interfaces::msg::Tag new_tag;
+      new_tag.id = tag_array_msg->tags[i].id;
+
+      geometry_msgs::msg::Point old_tag_position = tag_array_msg->tags[i].pose.position;
+
+      // extract the important parts of the pose
+      Eigen::Vector4d pose = Eigen::Vector4d(old_tag_position.x, old_tag_position.y, old_tag_position.z, 1);
+
+      // apply the transform
+      pose = eigen_transform.matrix() * optical * pose;
+
+      new_tag.pose.position.x = pose(0);
+      new_tag.pose.position.y = pose(1);
+      new_tag.pose.position.z = pose(2);
+
+      // get the orientation of the tag
+      Eigen::Matrix4d tag_orientation = convertQuatToRotMat(tag_array_msg->tags[i].pose.orientation);
+      tag_orientation = eigen_transform.matrix() * optical * tag_orientation;
+      new_tag.pose.orientation = convertRotMatToQuat(tag_orientation);
+
+      new_tag_array_msg.tags.push_back(new_tag);
+    }
+    tag_publisher_->publish(new_tag_array_msg);
+    // END STUDENT CODE
+
+    // visualization code
+    visualization_msgs::msg::MarkerArray marker_array;
+    for(stsl_interfaces::msg::Tag tag : new_tag_array_msg.tags)
+    {
+      visualization_msgs::msg::Marker marker;
+      marker.header.frame_id = "base_footprint";
+      marker.header.stamp = tag_array_msg->header.stamp;
+      marker.id = tag.id;
+      marker.type = visualization_msgs::msg::Marker::CUBE;
+      marker.action = visualization_msgs::msg::Marker::ADD;
+      marker.lifetime = rclcpp::Duration(0.25s);
+
+      marker.pose.orientation.x = tag.pose.orientation.x;
+      marker.pose.orientation.y = tag.pose.orientation.y;
+      marker.pose.orientation.z = tag.pose.orientation.z;
+      marker.pose.orientation.w = tag.pose.orientation.w;
+
+      marker.pose.position.x = tag.pose.position.x;
+      marker.pose.position.y = tag.pose.position.y;
+      marker.pose.position.z = tag.pose.position.z;
+
+      marker.scale.x = 0.2;
+      marker.scale.y = 0.2;
+      marker.scale.z = 0.025;
+
+      marker.color.a = 0.9;
+      marker.color.r = 0.0;
+      marker.color.b = 1.0;
+      marker.color.g = 0.0;
+
+      marker_array.markers.push_back(marker);
+    }
+    visualization_pub_->publish(marker_array);
+  }
+
+  Eigen::Matrix4d getRotMatForOpticalFrame() {
+    // BEGIN STUDENT CODE
+    Eigen::Matrix4d R_roll;
+    R_roll << 1, 0, 0, 0,
+            0, cos(M_PI/2), sin(M_PI/2), 0,
+            0, -sin(M_PI/2), cos(M_PI/2), 0,
+            0,0,0,1;
+
+    Eigen::Matrix4d R_yaw;
+    R_yaw << cos(M_PI/2), sin(M_PI/2), 0, 0,
+            -sin(M_PI/2), cos(M_PI/2), 0, 0,
+            0, 0, 1, 0,
+            0,0,0,1;
+    return R_yaw * R_roll;
+    // END STUDENT CODE
+  }
+
+  Eigen::Matrix4d convertQuatToRotMat(geometry_msgs::msg::Quaternion quat) {
+    Eigen::Quaterniond e_quat = Eigen::Quaterniond(quat.w, quat.x, quat.y, quat.z);
```

Comment:
Potential alternative. I don't feel strongly about this one. It is kinda nice to not have to worry about getting the ordering of the members correct.

```suggestion
    Eigen::Quaterniond e_quat;
    tf2::fromMsg(quat, e_quat);
```